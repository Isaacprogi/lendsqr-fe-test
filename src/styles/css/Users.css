.user-board {
  height: 100%;
  padding-right: 1rem;
  padding-left: 1rem;
  overflow-x: hidden;
  width: 100%;
}
@media screen and (min-width: 568px) {
  .user-board {
    padding-left: 2rem;
    padding-right: 2rem;
  }
}
.user-board h1 {
  font-family: "Work Sans";
  font-style: normal;
  font-weight: 500;
  font-size: 24px;
  color: #213F7D;
  margin-top: 3rem;
  margin-bottom: 3rem;
}
.user-board .status-board {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  gap: 1rem;
}
.user-board .status-board div {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  height: 10rem;
  background: white;
  border: 1px solid rgba(33, 63, 125, 0.06);
  -webkit-box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
          box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
  border-radius: 4px;
  height: 160px;
}
@media screen and (min-width: 450px) {
  .user-board .status-board div {
    -webkit-box-align: start;
        -ms-flex-align: start;
            align-items: flex-start;
    -webkit-box-pack: center;
        -ms-flex-pack: center;
            justify-content: center;
    padding-left: 2rem;
  }
}
.user-board .status-board div span.circle {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  background: #e9d5ec;
  cursor: pointer;
}
.user-board .status-board div .title {
  font-family: "Work Sans";
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 16px;
  text-transform: uppercase;
  color: #545F7D;
  margin-block: 1rem;
}
.user-board .status-board div .value {
  font-family: "Work Sans";
  font-style: normal;
  font-weight: 600;
  font-size: 24px;
  line-height: 28px;
  text-transform: uppercase;
  color: #213F7D;
}
@media screen and (min-width: 568px) {
  .user-board .status-board div {
    -ms-flex-item-align: start;
        align-self: flex-start;
    -webkit-box-pack: center;
        -ms-flex-pack: center;
            justify-content: center;
  }
}
.user-board .table-container {
  width: 100%;
  height: 750px;
  margin-top: 1rem;
  border: 1px solid rgba(33, 63, 125, 0.06);
  -webkit-box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
          box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
  border-radius: 4px;
  position: relative;
}
.user-board .table-container .spinner {
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%, -50%);
          transform: translate(-50%, -50%);
}
.user-board .table-container .table {
  width: 100%;
}
.user-board .table-container .table .table-titles {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
}
.user-board .table-container .table .table-titles span {
  -webkit-box-flex: 1;
      -ms-flex: 1;
          flex: 1;
  padding: 0.7rem;
  min-width: 6rem;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: start;
      -ms-flex-pack: start;
          justify-content: flex-start;
}
.user-board .table-container .table .table-titles span img {
  margin-left: 0.5rem;
  height: 10.67px;
  width: 16px;
}
.user-board .table-container .table .table-titles .email, .user-board .table-container .table .table-titles .date-joined, .user-board .table-container .table .table-titles .phone-number, .user-board .table-container .table .table-titles .organization {
  display: none;
}
@media screen and (min-width: 450px) {
  .user-board .table-container .table .table-titles .email {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
  }
}
@media screen and (min-width: 988px) {
  .user-board .table-container .table .table-titles .phone-number, .user-board .table-container .table .table-titles .organization {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
  }
}
@media screen and (min-width: 1200px) {
  .user-board .table-container .table .table-titles .date-joined {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
  }
}
.user-board .table-container .table .table-data-container {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
}
.user-board .table-container .table .table-data-container .table-data {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}
.user-board .table-container .table .table-data-container .table-data .cover {
  width: 93%;
  height: 100%;
  position: absolute;
  cursor: pointer;
}
.user-board .table-container .table .table-data-container .table-data span {
  -webkit-box-flex: 1;
      -ms-flex: 1;
          flex: 1;
  padding: 0.7rem;
  min-width: 6rem;
  word-break: break-all;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: start;
      -ms-flex-align: start;
          align-items: flex-start;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
  position: relative;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) .status-options {
  position: absolute;
  background-color: white;
  display: none;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  border-radius: 0.4rem;
  z-index: 400;
  border: 1px solid rgba(84, 95, 125, 0.04);
  -webkit-box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
          box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
  border-radius: 4px;
  width: 12rem;
  top: 0;
  right: 2.6rem;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) .status-options span {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: start;
      -ms-flex-pack: start;
          justify-content: flex-start;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) .status-options span:hover {
  background-color: #e3f3f3;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) .status-options span img {
  cursor: pointer;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) .visible {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) img {
  margin-right: 0.6rem;
  cursor: pointer;
  margin-left: 0.5rem;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) div {
  text-overflow: ellipsis;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) div.Inactive {
  color: rgb(84, 95, 125);
  background: rgba(84, 95, 125, 0.06);
  border-radius: 100px;
  padding-left: 0.7rem;
  padding-right: 0.7rem;
  padding-top: 0.4rem;
  padding-bottom: 0.4rem;
  overflow: hidden;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) div.Active {
  color: rgb(57, 205, 98);
  background: rgba(57, 205, 98, 0.06);
  border-radius: 100px;
  padding-left: 0.7rem;
  padding-right: 0.7rem;
  padding-top: 0.4rem;
  padding-bottom: 0.4rem;
  overflow: hidden;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) div.Blacklisted {
  color: rgb(228, 3, 59);
  background: rgba(228, 3, 59, 0.1);
  border-radius: 100px;
  padding-left: 0.7rem;
  padding-right: 0.7rem;
  padding-top: 0.4rem;
  padding-bottom: 0.4rem;
  overflow: hidden;
}
.user-board .table-container .table .table-data-container .table-data :nth-last-child(1) div.Pending {
  color: rgb(233, 178, 0);
  background: rgba(233, 178, 0, 0.1);
  border-radius: 100px;
  padding-left: 0.7rem;
  padding-right: 0.7rem;
  padding-top: 0.4rem;
  padding-bottom: 0.4rem;
  overflow: hidden;
}
.user-board .table-container .table .table-data-container .table-data .email, .user-board .table-container .table .table-data-container .table-data .date-joined, .user-board .table-container .table .table-data-container .table-data .phone-number, .user-board .table-container .table .table-data-container .table-data .organization {
  display: none;
}
@media screen and (min-width: 450px) {
  .user-board .table-container .table .table-data-container .table-data .email {
    display: block;
  }
}
@media screen and (min-width: 988px) {
  .user-board .table-container .table .table-data-container .table-data .phone-number, .user-board .table-container .table .table-data-container .table-data .organization {
    display: block;
  }
}
@media screen and (min-width: 1200px) {
  .user-board .table-container .table .table-data-container .table-data .date-joined {
    display: block;
  }
}
.user-board .pagination {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  -ms-flex-wrap: wrap;
      flex-wrap: wrap;
  margin-top: 1rem;
  font-family: "Work Sans";
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  color: #545F7D;
  padding-bottom: 2rem;
}
.user-board .pagination .drop-down {
  width: 270px;
  height: 600px;
  position: absolute;
  top: -605px;
  z-index: 400;
  background: white;
  border: 1px solid rgba(84, 95, 125, 0.14);
  -webkit-box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
          box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
  border-radius: 4px;
  left: 0.6rem;
  padding-left: 0.8rem;
  padding-right: 0.8rem;
  padding-block: 1rem;
  -webkit-transform: scale(0);
          transform: scale(0);
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}
.user-board .pagination .drop-down div {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  background-color: white;
  color: #545F7D;
}
.user-board .pagination .drop-down div span {
  margin-top: 1rem;
}
.user-board .pagination .drop-down div select, .user-board .pagination .drop-down div input {
  padding: 0.6rem;
  background: white;
  border: 1px solid rgba(33, 63, 125, 0.2);
  border-radius: 8px;
  margin-top: 0.5rem;
  outline: none;
  width: 100%;
}
.user-board .pagination .drop-down div select > option {
  padding: 0.2rem !important;
  background-color: rgb(163, 171, 194) !important;
}
.user-board .pagination .drop-down .controls {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
  width: 100%;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  margin-top: 1rem;
}
.user-board .pagination .drop-down .controls :first-child {
  color: #545F7D;
  border: 1px solid #545F7D;
  border-radius: 8px;
}
.user-board .pagination .drop-down .controls :last-child {
  color: white;
  margin-left: 1rem;
  background: #39CDCC;
  border: 1px solid #39CDCC;
  border-radius: 8px;
}
.user-board .pagination .drop-down .controls span {
  width: 98px;
  height: 40px;
  background: white;
  border: 1px solid rgba(84, 95, 125, 0.14);
  -webkit-box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
          box-shadow: 3px 5px 20px rgba(0, 0, 0, 0.04);
  border-radius: 4px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  cursor: pointer;
}
.user-board .pagination .drop-down.visible {
  -webkit-transform: scale(1);
          transform: scale(1);
  -webkit-transform-origin: left;
          transform-origin: left;
}
.user-board .pagination .right {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: start;
      -ms-flex-pack: start;
          justify-content: flex-start;
  margin-bottom: 0.5rem;
}
@media screen and (min-width: 768px) {
  .user-board .pagination .right {
    margin-bottom: 0;
  }
}
.user-board .pagination .right span {
  margin-right: 0.4rem;
  color: rgb(84, 95, 125);
}
.user-board .pagination .right .button-chevron-left {
  rotate: 45deg;
  width: 100px;
}
.user-board .pagination .right > :nth-last-child(2) {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  background: rgba(33, 63, 125, 0.1);
  border-radius: 4px;
  padding: 0.3rem;
  color: #213F7D;
  cursor: pointer;
}
.user-board .pagination .right > :nth-last-child(2) img {
  margin-left: 0.7rem;
}
.user-board .pagination .left .pagination-left {
  list-style: none;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
  font-size: 1.2rem;
  gap: 5px;
}
.user-board .pagination .left .pagination-left .page-num {
  padding: 0 5px;
  cursor: pointer;
  border-radius: 3px;
  font-weight: 400;
}
.user-board .pagination .left .pagination-left .page-num:hover {
  background-color: #6d7686;
  color: white;
}
.user-board .pagination .left .pagination-left .page-nav-button-1 {
  position: relative;
  background: rgba(33, 63, 125, 0.1);
  border-radius: 4px;
  padding: 0.1rem;
  font-size: 1rem;
  cursor: pointer;
  width: 1.4rem;
  height: 1.4rem;
  margin-right: 0.5rem;
}
.user-board .pagination .left .pagination-left .page-nav-button-1:hover {
  background-color: #9fa8b7;
  color: white;
}
.user-board .pagination .left .pagination-left .page-nav-button-2 {
  background: rgba(33, 63, 125, 0.1);
  border-radius: 4px;
  padding: 0.1rem;
  font-size: 1rem;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  cursor: pointer;
  width: 1.4rem;
  height: 1.4rem;
}
.user-board .pagination .left .pagination-left .page-nav-button-2:hover {
  background-color: #9fa8b7;
  color: white;
}
.user-board .pagination .left .pagination-left .active {
  background-color: #213F7D;
}
@media screen and (min-width: 768px) {
  .user-board .pagination {
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
        -ms-flex-direction: row;
            flex-direction: row;
    -webkit-box-pack: justify;
        -ms-flex-pack: justify;
            justify-content: space-between;
    -webkit-box-align: center;
        -ms-flex-align: center;
            align-items: center;
  }
}/*# sourceMappingURL=Users.css.map */