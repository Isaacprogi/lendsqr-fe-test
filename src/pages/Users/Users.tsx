import { useEffect, useRef, useState } from "react";
import UsersLogo from '../../assets/svg/user-activities/Users.svg';
import ActiveUsers from '../../assets/svg/user-activities/ActiveUsers.svg';
import UserWithLoans from '../../assets/svg/user-activities/UserWithLoans.svg';
import UserWithSavings from '../../assets/svg/user-activities/UserWithSavings.svg';
import SignalDrop from '../../assets/svg/user-activities/SignalDrop.svg';
import { BsChevronLeft } from "react-icons/bs"
import { BsChevronRight } from "react-icons/bs"
import chevronDown from '../../assets/svg/user-activities/chevronDown.svg';
import { UserData } from "../../components/UserData/UserData";
import '../../styles/css/Users.css';
import ReactPaginate from 'react-paginate';
import { StatusBoardItem } from "../../components/StatusBoardItem/StatusBoardItem";
import { userDataFetch } from "../../utils/types/interface_and_types";
import axios from 'axios';

export const Users = () => {
    const [users, setUsers] = useState<userDataFetch[]>([]);
    const n = 1;

    const [selectData, setSelectData] = useState<string>('');
    const [email, setEmail] = useState<string>('');
    const [date, setDate] = useState<string>('');
    const [statusData, setStatusData] = useState<string>('');
    const [phoneNumber, setPhoneNumber] = useState<string>('');
    const [username, setUsername] = useState<string>('');
    const [organizations, setOrganizations] = useState<string[]>([]);
    const [loading, setLoading] = useState(false);
    const [displayStatusOptions, setDisplayStatusOptions] = useState<string>('');
    const [filter, setFilter] = useState<userDataFetch[]>([]);

    const [dropDownActive, setDropDownActive] = useState(false);
    const dropdownRef = useRef<HTMLDivElement>(null);

    const randomStatus = () => {
        return ['Active', 'Inactive', 'Blacklisted', 'Pending']?.sort(() => Math.random() > 0.5 ? 1 : -1)?.slice(0, n);
    };

    const getUsers = async () => {
        setLoading(true);
        try {
            const response = await axios.get<userDataFetch[]>('https://6270020422c706a0ae70b72c.mockapi.io/lendsqr/api/v1/users');
            const newData = response.data.map((item: any) => {
                return { ...item, status: randomStatus().toString() };
            });
            setUsers(newData);
            setFilter(newData);
            setOrganizations([...new Set(response.data.map((item: any) => item?.orgName))]);
            setLoading(false);
        } catch (error) {
            setLoading(false);
            console.log('Error fetching data:', error);
        }
    };

    useEffect(() => {
        getUsers();
    }, []);

    const handleFilter = () => {
        setFilter(users.filter((user: userDataFetch) => user?.userName === username
            || user.orgName === selectData
            || new Date(user.createdAt).toDateString() === new Date(date).toDateString()
            || user.status.toLowerCase() === statusData.toLowerCase()
            || user.phoneNumber === phoneNumber
            || user.email === email
        ));
    };

    const handleReset = () => {
        setSelectData("");
        setUsername("");
        setDate("");
        setEmail("");
        setStatusData("");
        setPhoneNumber("");
        getUsers();
    };

    const [currentItems, setCurrentItems] = useState<any[]>([]);
    const [pageCount, setPageCount] = useState(0);
    const [itemOffset, setItemOffset] = useState(0);
    const itemsPerPage = 9;

    useEffect(() => {
        const endOffset = itemOffset + itemsPerPage;
        setCurrentItems(filter.slice(itemOffset, endOffset));
        setPageCount(Math.ceil(filter.length / itemsPerPage));
    }, [itemOffset, itemsPerPage, filter]);

    const handlePageClick = (event: any) => {
        const newOffset = (event.selected * itemsPerPage) % filter.length;
        setItemOffset(newOffset);
    };

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
                setDropDownActive(false);
            }
        };
        document.addEventListener("mousedown", handleClickOutside);
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, [dropdownRef]);

    return loading === false && users?.length > 0 ? (
        <div className="user-board">
            <h1>Users</h1>

            <section className="status-board">
                <StatusBoardItem logo={UsersLogo} title="Users" value={users.length.toString()} />
                <StatusBoardItem logo={ActiveUsers} title="Active Users" value={users.filter(user => user.status === 'Active').length.toString()} />
                <StatusBoardItem logo={UserWithLoans} title="User With Loans" value={users.filter(user => user.education?.loanRepayment !== '').length.toString()} />
                <StatusBoardItem logo={UserWithSavings} title="User With Savings" value={users.filter(user => user.accountBalance !== '').length.toString()} />
            </section>

            <section className="table-container">
                <div className="table">
                    <section className="table-titles">
                        <span className="organization">
                            <span data-testid="table-title-1">Organization</span>
                            <img src={SignalDrop} alt="organization-logo" />
                        </span>
                        <span className="username">
                            <span data-testid="table-title-2">Username</span>
                            <img src={SignalDrop} alt="username-logo" />
                        </span>
                        <span className="email">
                            <span data-testid="table-title-3">Email</span>
                            <img src={SignalDrop} alt="email-logo" />
                        </span>
                        <span className="phone-number">
                            <span data-testid="table-title-4">Phone Number</span>
                            <img src={SignalDrop} alt="phone-number-logo" />
                        </span>
                        <span className="date-joined">
                            <span data-testid="table-title-5">Date</span>
                            <img src={SignalDrop} alt="date-logo" />
                        </span>
                        <span className="status">
                            <span data-testid="table-title-6">Status</span>
                            <img src={SignalDrop} alt="status-logo" />
                        </span>
                    </section>
                    <section className="table-data-container">
                        {
                            !loading && currentItems.map(user => {
                                return <UserData displayStatusOptions={displayStatusOptions} setDisplayStatusOptions={setDisplayStatusOptions} key={user.id} user={user} />;
                            })
                        }
                    </section>
                </div>
            </section>

            <section className="pagination">
                <div data-testid="drop-down-block" ref={dropdownRef} className={`drop-down ${dropDownActive ? 'visible' : ''}`}>
                    <div className="select">
                        <span data-testid="select-organization">Organization</span>
                        <select name="select-input" value={selectData} onChange={(e) => setSelectData(e.target.value)}>
                            {
                                organizations?.map((org, index) => {
                                    return <option key={index}>
                                        {org}
                                    </option>;
                                })
                            }
                        </select>
                    </div>
                    <div className="username">
                        <span data-testid="username">Username</span>
                        <input value={username} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setUsername(e.target.value)} placeholder="Username" type="text" />
                    </div>
                    <div className="email">
                        <span data-testid="email">Email</span>
                        <input value={email} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setEmail(e.target.value)} placeholder="Email" type="email" />
                    </div>
                    <div className="date">
                        <span data-testid="date">Date</span>
                        <input value={date} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setDate(e.target.value)} type="date" />
                    </div>
                    <div className="phone-number">
                        <span data-testid="phone-number">Phone Number</span>
                        <input value={phoneNumber} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setPhoneNumber(e.target.value)} placeholder="Phone Number" type="text" />
                    </div>
                    <div className="status">
                        <span data-testid="status">Status</span>
                        <input value={statusData} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setStatusData(e.target.value)} placeholder="Status" type="text" />
                    </div>

                    <div className="controls">
                        <span onClick={handleReset} data-testid="reset">Reset</span>
                        <span onClick={handleFilter} data-testid="filter">Filter</span>
                    </div>
                </div>
                <div className="right">
                    <span>Showing</span>
                    <span data-testid="drop-down-button" onClick={() => setDropDownActive(!dropDownActive)}>
                        {filter?.length}
                        <img src={chevronDown} alt="chevron-down" />
                    </span>
                    <span>out of 100</span>
                </div>
                <div className="left">
                    <ReactPaginate
                        breakLabel="..."
                        nextLabel={<BsChevronRight />}
                        onPageChange={handlePageClick}
                        pageRangeDisplayed={3}
                        pageCount={pageCount}
                        previousLabel={<BsChevronLeft />}
                        containerClassName="pagination-left"
                        pageLinkClassName="page-num"
                        previousClassName="page-nav-button-1"
                        nextLinkClassName="page-nav-button-2"
                        activeClassName="active"
                    />
                </div>
            </section>
        </div>
    ) : 
        <div className="skeleton">
            <div className="skeleton-line-1"></div>
            <div className="skeleton-line-2"></div>
            <div className="skeleton-line-3"></div>
        </div>
};
