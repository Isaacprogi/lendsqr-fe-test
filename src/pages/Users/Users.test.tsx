import { render, screen, waitFor, act, fireEvent } from '@testing-library/react';
import { describe, test, vi, beforeEach, afterEach } from 'vitest';
import { Users } from './Users';
import { BrowserRouter } from 'react-router-dom';
import { userMockData } from '../../ts/mockdata';
import axios from 'axios';
import renderer from 'react-test-renderer';

vi.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('Users Component Tests', () => {
    beforeEach(() => {
        mockedAxios.get.mockResolvedValueOnce({ data: userMockData });
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    test('data should be present after API call is made', async () => {
        await act(async () => {
            render(
                <BrowserRouter>
                    <Users />
                </BrowserRouter>
            );
        });

        expect(axios.get).toHaveBeenCalledWith(`https://6270020422c706a0ae70b72c.mockapi.io/lendsqr/api/v1/users`);
        expect(axios.get).toHaveBeenCalledTimes(1);

        await waitFor(() => {
            const divElements = screen.getAllByTestId('user-data-box');
            expect(divElements.length).toBe(userMockData.length);
            expect(divElements[0].innerHTML).toContain(`<span class="organization">${userMockData[0].orgName}</span>`);
        });
    });

    test('should match snapshot', () => {
        const tree = renderer.create(
            <BrowserRouter>
                <Users />
            </BrowserRouter>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should filter users based on filter inputs', async () => {
        await act(async () => {
            render(
                <BrowserRouter>
                    <Users />
                </BrowserRouter>
            );
        });

        const usernameInput = screen.getByPlaceholderText('Username');
        fireEvent.change(usernameInput, { target: { value: userMockData[0].userName } });

        const filterButton = screen.getByTestId('filter');
        fireEvent.click(filterButton);

        await waitFor(() => {
            const divElements = screen.getAllByTestId('user-data-box');
            expect(divElements.length).toBe(2);
            expect(divElements[0].innerHTML).toContain(`<span class="organization">${userMockData[0].orgName}</span>`);
        });
    });

});
