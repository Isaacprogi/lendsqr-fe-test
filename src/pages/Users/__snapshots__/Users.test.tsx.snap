// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Users Component Tests > should match snapshot 1`] = `
<div
  className="skeleton"
>
  <div
    className="skeleton-line-1"
  />
  <div
    className="skeleton-line-2"
  />
  <div
    className="skeleton-line-3"
  />
</div>
`;
